observers/qos_correction.hpp

```diff
@@ -0,0 +1,132 @@
+#ifndef SERENITY_QOS_CORRECTION_HPP
+#define SERENITY_QOS_CORRECTION_HPP
+
+#include <list>
+#include <vector>
+
+#include "glog/logging.h"
+
+#include "mesos/mesos.hpp"
+#include "mesos/resources.hpp"
+
+#include "mesos/slave/oversubscription.hpp"
+
+#include "messages/serenity.hpp"
+
+#include "serenity/serenity.hpp"
+
+namespace mesos {
+namespace serenity {
+
+inline std::list<ResourceUsage_Executor> filterPrExecutors(
+    ResourceUsage usage) {
+  std::list<ResourceUsage_Executor> beExecutors;
+  for (ResourceUsage_Executor inExec : usage.executors()) {
+    if (!inExec.has_executor_info()) {
+      LOG(ERROR) << "Executor <unknown>"
+      << " does not include executor_info";
+      // Filter out these executors.
+      continue;
+    }
+    if (inExec.allocated().size() == 0) {
+      LOG(ERROR) << "Executor "
+      << inExec.executor_info().executor_id().value()
+      << " does not include allocated resources.";
+      // Filter out these executors.
+      continue;
+    }
+
+    Resources allocated(inExec.allocated());
+    // Check if task uses revocable resources.
+    if (!allocated.revocable().empty())
+      beExecutors.push_back(inExec);
+  }
+
+  return beExecutors;
+}
+
+
+using ContentionInterpreterFunction = Try<QoSCorrections>
+    (Contentions currentContentions,
+    ResourceUsage currentUsage);
+
+
+/**
+ * Class including all algorithms for contention interpretations.
+ * They converts contentions & usage to QoSCorrections.
+ *
+ * Convenient for debuging and testing different algorithms.
+ */
+class ContentionInterpreters {
+ public:
+  static ContentionInterpreterFunction severityBasedCpuContention;
+};
+
+
+/**
+ * QoSCorrectionObserver observes incoming Contentions and
+ * ResourceUsage and produces QoSCorrections.
+ *
+ * Currently Slave understand only kill correction action.
+ */
+class QoSCorrectionObserver : public MultiConsumer<Contentions>,
+                              public Consumer<ResourceUsage>,
+                              public Producer<QoSCorrections> {
+ public:
+  explicit QoSCorrectionObserver(
+      Consumer<QoSCorrections>* _consumer,
+      uint64_t _contentionProducents,
```

Comment:
Number of contentions products we are expecting in our consumer. (It's filling MultiConsumer(..) constructor)